### PR TITLE
OIDC support on the CLI

### PR DESCRIPTION
**Pull request by maraino**
### Description
This PR makes the CLI compatible with the provisioner refactoring in certificates and adds support for OIDC tokens in the ca commands.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -64,7 +64,7 @@ required = [
   unused-packages = true
 
 [[constraint]]
-  branch = "master"
+  branch = "oidc-provisioner"
```
Comment (dopey) [inline Gopkg.toml]: let's just remember to change this. I always feel like I'll forget.

File: cmd/step/main.go [hunk 2]
```diff
@@ -93,6 +95,22 @@ func main() {
...
+func panicHandler() {
+	if r := recover(); r != nil {
+		if os.Getenv("STEPDEBUG") == "1" {
+			fmt.Fprintf(os.Stderr, "%s\n", config.Version())
+			fmt.Fprintf(os.Stderr, "Release Date: %s\n\n", config.ReleaseDate())
+			panic(r)
+		} else {
+			fmt.Fprintln(os.Stderr, "Something unexpected happened.")
+			fmt.Fprintln(os.Stderr, "If you want to help us debug the problem, please run:")
+			fmt.Fprintf(os.Stderr, "STEPDEBUG=1 %s\n", strings.Join(os.Args, " "))
```
Comment (dopey) [inline cmd/step/main.go]: nice!

File: Gopkg.toml (hunk 1 above)
Comment (maraino) [reply Gopkg.toml]: 👍

File: command/ca/certificate.go [hunk 3]
```diff
@@ -302,3 +343,69 @@ func (f *certificateFlow) Sign(ctx *cli.Context, token string, csr api.Certifica
...
+		return nil, nil, errors.Wrap(err, "error parsing token")
+	}
+
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error generating key")
+	}
+
+	var emails []string
+	dnsNames, ips := splitSANs(sans, claims.SANs)
```
Comment (dopey) [inline command/ca/certificate.go]: should we be parsing emails out in splitSANS as well?

File: command/ca/certificate.go [hunk 4]
```diff
@@ -302,3 +343,69 @@ func (f *certificateFlow) Sign(ctx *cli.Context, token string, csr api.Certifica
...
+	tok, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error parsing token")
+	}
+	var claims tokenClaims
+	if err := tok.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		return nil, nil, errors.Wrap(err, "error parsing token")
+	}
+
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```
Comment (dopey) [inline command/ca/certificate.go]: Any reason we're using ecdsa explicitly instead of using the keys.Generate... which uses the defaults?

Comment (maraino) [reply command/ca/certificate.go]: I copied the code from the CA, and it was like that, I'll change it.

File: command/ca/certificate.go (hunk 3 above)
Comment (maraino) [reply command/ca/certificate.go]: At some point, it would be a new feature and it's independent of this PR.

File: command/ca/provisioner/add.go [hunk 5]
```diff
@@ -74,104 +137,167 @@ func addAction(ctx *cli.Context) (err error) {
...
 	if err != nil {
 		return errors.Wrapf(err, "error loading configuration")
 	}
 
-	var provisioners []*authority.Provisioner
-	provMap := make(map[string]*authority.Provisioner)
-	for _, prov := range c.AuthorityConfig.Provisioners {
-		provisioners = append(provisioners, prov)
-		provMap[prov.Name+":"+prov.Key.KeyID] = prov
+	typ := strings.ToUpper(ctx.String("type"))
```
Comment (dopey) [inline command/ca/provisioner/add.go]: in the docs you say this is case-sensitive, but it doesn't appear to be.

Comment (maraino) [reply command/ca/provisioner/add.go]: It says `Type is a case-insensitive string`

Comment (dopey) [reply command/ca/provisioner/add.go]: I'm bad at reading.

File: command/ca/provisioner/remove.go [hunk 6]
```diff
@@ -21,7 +22,11 @@ func removeCommand() cli.Command {
 			},
 			cli.StringFlag{
 				Name:  "kid",
-				Usage: "The <kid> (Key ID) of for the provisioner key to be removed.",
+				Usage: "The <kid> (Key ID) of for the JWK provisioner key to be removed.",
```
Comment (dopey) [inline command/ca/provisioner/remove.go]: The <kid> (Key ID) of the JWK provisioner key to be removed.

Comment (dopey) [reply command/ca/provisioner/remove.go]: "of for" fix

File: command/ca/provisioner/remove.go [hunk 7]
```diff
@@ -21,7 +22,11 @@ func removeCommand() cli.Command {
...
+			},
+			cli.StringFlag{
+				Name:  "client-id",
+				Usage: "The <id> (Client ID) of for the OIDC provisioner to be removed.",
```
Comment (dopey) [inline command/ca/provisioner/remove.go]: "of for" fix

Comment (dopey) [inline command/ca/provisioner/remove.go]: --all flag also cannot be used in conjunction with the --client-id flag. (update docs for --all)

**Pull request merged**